File: src/leetcode/p051to100/LeetCode084LargestRectangleInHistogram.java
package leetcode.p051to100;

import java.util.Stack;

public class LeetCode084LargestRectangleInHistogram {

    private interface Method {
        int largestRectangleArea(int[] heights);
    }

    private static final class Greedy implements Method {

        public int largestRectangleArea(int[] heights) {
            if (heights == null || heights.length == 0) {
                return 0;
            }
            int max = 0;
            Stack<Integer> stack = new Stack<>();
            for (int i = 0; i <= heights.length; ) {
                if (stack.isEmpty() || i < heights.length && heights[i] >= heights[stack.peek()]) {
                    // The design is that we push i into stack.
                    // And only pop the stack when the cursor move to j, where heights[i] > heights[j]
                    // In this case, we find that all heights between i and j are not lower than height[i]
                    stack.push(i++);
                } else {
                    int hIndex = stack.pop();
                    int height = heights[hIndex];
                    int length = stack.isEmpty() ? i : i - 1 - stack.peek();
                    // area = height x length
                    // We pop and get a hIndex.
                    // hIndex may be discontiguous to stack.peek(),
                    // because there might be an n (stack.peek() < n && n < hIndex), heights[n] > heights[hIndex],
                    // and n has already been popped.
                    // Also, n (hIndex < n && i < n) must have heights[n] >= heights[hIndex].
                    // All in all, for hIndex, stack.peek() is the first index whose height is lower than hIndex,
                    // i is the last index whose height is lower than hIndex.
                    // [stack.peek() + 1, i - 1] is the range of rectangle whose height are not lower than hIndex.
                    int area = height * length;
                    max = Math.max(max, area);
                }
            }
            return max;
        }
    }
}

File: src/leetcode/p051to100/LeetCode084LargestRectangleInHistogram.kt
package leetcode.p051to100

import java.util.Stack

class LeetCode084LargestRectangleInHistogram {

    fun largestRectangleArea(heights: IntArray?): Int {
        if (heights == null || heights.isEmpty()) {
            return 0
        }
        var maxArea = 0
        val stack = Stack<Int>()
        var i = 0
        while (i <= heights.size) {
            if (stack.isEmpty() || i < heights.size && heights[i] >= heights[stack.peek()]) {
                stack.push(i++)
            } else {
                val hIndex = stack.pop()
                // area = height x length
                // Because we use the height in hIndex rather than in i, (i - hIndex) isn't the length.
                // The key is stack.peek() it store the first index that is shorter than heights[hIndex]
                // Because hIndex can be contiguous or discontiguous to i,
                // the right boundary should be (i - 1).
                // The length == (i - 1) - stack.peek()
                val area = heights[hIndex] * if (stack.isEmpty()) i else i - 1 - stack.peek()
                maxArea = Math.max(maxArea, area)
            }
        }
        return maxArea
    }

    companion object {

        @JvmStatic
        fun main(args: Array<String>) {
            val one = LeetCode084LargestRectangleInHistogram()
            val heights = intArrayOf(4, 6, 5, 7, 8, 3)
            one.largestRectangleArea(heights)
        }
    }

}
